Examples/Io/EDM4hep/src/EDM4hepUtil.cpp

```diff
@@ -0,0 +1,305 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2022 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "ActsExamples/Io/EDM4hep/EDM4hepUtil.hpp"
+
+#include "Acts/Definitions/Units.hpp"
+#include "Acts/EventData/MultiTrajectory.hpp"
+#include "Acts/EventData/MultiTrajectoryHelpers.hpp"
+#include "Acts/Utilities/Helpers.hpp"
+#include "ActsExamples/Digitization/MeasurementCreation.hpp"
+#include "ActsExamples/EventData/Index.hpp"
+#include "ActsExamples/EventData/IndexSourceLink.hpp"
+#include "ActsExamples/EventData/SimHit.hpp"
+#include "ActsExamples/Validation/TrackClassification.hpp"
+
+#include "edm4hep/TrackState.h"
+
+namespace ActsExamples {
+
+ActsFatras::Particle EDM4hepUtil::readParticle(
+    edm4hep::MCParticle from, MapParticleIdFrom particleMapper) {
+  ActsFatras::Barcode particleId = particleMapper(from);
+
+  ActsFatras::Particle to(particleId, (Acts::PdgParticle)from.getPDG(),
+                          from.getCharge() * Acts::UnitConstants::e,
+                          from.getMass() * Acts::UnitConstants::GeV);
+
+  // TODO do we have that in EDM4hep?
+  // particle.setProcess(static_cast<ActsFatras::ProcessType>(data.process));
+
+  to.setPosition4(from.getVertex()[0] * Acts::UnitConstants::mm,
+                  from.getVertex()[1] * Acts::UnitConstants::mm,
+                  from.getVertex()[2] * Acts::UnitConstants::mm,
+                  from.getTime() * Acts::UnitConstants::ns);
+
+  // Only used for direction; normalization/units do not matter
+  to.setDirection(from.getMomentum()[0], from.getMomentum()[1],
+                  from.getMomentum()[2]);
+
+  to.setAbsoluteMomentum(std::hypot(from.getMomentum()[0],
+                                    from.getMomentum()[1],
+                                    from.getMomentum()[2]) *
+                         Acts::UnitConstants::GeV);
+
+  return to;
+}
+
+void EDM4hepUtil::writeParticle(const ActsFatras::Particle& from,
+                                edm4hep::MutableMCParticle to) {
+  // TODO what about particleId?
```

Comment:
So we can't writer our barcode in there? Maybe an input to edm4hep as well: add a user defined particle identifier?

Reply:
yeah exactly they don't offer a field for particle ID. it is only identified by an index in their collection

Reply:
Ok this is something we should clarify with the edm4hep people. They store the vertex location, so during reading we might have to reconstruct this.